src/server/server.zig

```diff
@@ -21,14 +24,59 @@ pub const storage = @import("storage.zig");
 const command = @import("command/_command.zig");
 
 pub const WorldEditData = struct {
+	const maxWorldEditHistoryCapacity: u32 = 1024;
+
 	selectionPosition1: ?Vec3i = null,
 	selectionPosition2: ?Vec3i = null,
-	clipboard: ?main.blueprint.Blueprint = null,
+	clipboard: ?Blueprint = null,
+	undoHistory: History,
+
+	const History = struct {
+		changes: CircularBufferQueue(Value),
 
+		const Value = struct {
+			blueprint: Blueprint,
+			position: Vec3i,
+			message: []const u8,
+
+			pub fn init(blueprint: Blueprint, position: Vec3i, message: []const u8) Value {
+				return .{.blueprint = blueprint, .position = position, .message = main.globalAllocator.dupe(u8, message)};
+			}
+			pub fn deinit(self: Value) void {
+				main.globalAllocator.free(self.message);
+				self.blueprint.deinit(main.globalAllocator);
+			}
+		};
+		pub fn init() History {
+			return .{.changes = .init(main.globalAllocator, maxWorldEditHistoryCapacity)};
+		}
+		pub fn deinit(self: *History) void {
+			self.clear();
+			self.changes.deinit();
+		}
+		pub fn clear(self: *History) void {
+			while(self.changes.dequeue()) |item| item.deinit();
+			self.changes.clearRetainingCapacity();
+		}
+		pub fn push(self: *History, value: Value) void {
+			self.changes.enqueue(value);
+
+			if(self.changes.full()) {
+				if(self.changes.dequeue()) |oldValue| oldValue.deinit();
+			}
```

Comment:
Should swap them to actually make use of the full capacity.
```suggestion
			if(self.changes.full()) {
				if(self.changes.dequeue()) |oldValue| oldValue.deinit();
			}

			self.changes.enqueue(value);
```